### ISSUE TITLE
Configuration option to retain source comments in wallaby instrumentation step

### ISSUE THREAD

**Issue by sambernet** [wallabyjs/public#1883]
### Background/Scenario

We recently acquired company licenses for a few devs. They now started using it on a range of existing projects. One is an AngularJS project that is typescript-based and uses the [webpack-angular-translate NPM library](https://www.npmjs.com/package/webpack-angular-translate) to generate default translation/localization files at build time.

This plugin uses some predefined comment tokens to whitelist otherwise disallowed usages of the translation calls with non-static translation keys (see [service docs](https://www.npmjs.com/package/webpack-angular-translate#service) for details). To mark a specific call explicitely as _safe_ (because translations have been manually registered for a given key), DEVs just add a `/* suppress-dynamic-translation-error: true */` comment before the offending call.

Now due to the fact that wallaby needs to perform the Typescript compilation/instrumentation before running the actual webpack build, those comments are removed from the compiled sources and the webpack build will fail.

We ensured that `tsconfig.json` is properly setup to keep the comments and verified that is actually the case. It looks like right now the comments are stripped out by the (non-transparent) **instrumentation step**.

We did adapt the webpack config for wallaby to make sure the pre-compiled/instrumented `.js` files will be processed with the `webpack-angular-translate` plugin which works fine. But due to the comments being removed, we have errors in the webpack build that we cannot remove.

#### other use-cases

Our issue above is one specific use-case to demonstrate why we would like to be able to keep comments on the original statement. 

I reckon there is a load of similar use cases - wherever some build-time tool can be configured using source code comments. 

A common example would be linting, where (false-positive) occurrences of linting errors can be whitelisted using comments. See [ESLint Configuration Comments](https://eslint.org/docs/user-guide/configuring#using-configuration-comments) as an example case. 

I guess most users now just resort to disabling the linting altogether in specific wallabyjs webpack configs. But besides the effort to figure out the problem and configure the wallabyjs webpack config to mitigate, this also means that after a wallabyJS session, one should always run the "real build" as well to ensure no linting issue sneaked into the current build.

### Question/Feature Request

Is it somehow possible to configure Wallaby to not remove source code comments in the instrumentation step?

### Wallaby.js configuration file

```javascript
var wallabyWebpack = require('wallaby-webpack'),
    webpackConfig = require('./webpack.config'),
    WebPackAngularTranslate = require("webpack-angular-translate");

module.exports = function (wallaby) {

    webpackConfig.module.rules.push({
        // wallaby compiles the TS files before starting webpack compilation in order to be able to instrument the code
        // thus we need to run the webpack-angular-translate loader also on JS files that have already been compiled.
        test: /((\\src\\)|(\/src\/)).*\.js$/,
        use: [{
            loader: WebPackAngularTranslate.jsLoader()
        }]
    });
    webpackConfig.entryPatterns = [
        'src/test-env.js',
        'src/**/*spec.js'
    ];
    var wallabyPostprocessor = wallabyWebpack(webpackConfig);

    return {
        files: [
            {pattern: 'src/**/*.+(ts|css|less|scss|sass|html|json|svg)', load: false},
            {pattern: 'src/**/*.spec.ts', ignore: true}
        ],
        tests: [
            {pattern: 'src/**/*.spec.ts', load: false}
        ],
        env: {
            kind: 'chrome'
        },
        testFramework: 'jasmine',
        postprocessor: wallabyPostprocessor,
        setup: function () {
            // required to trigger test loading
            window.__moduleBundler.loadTests();
        },
        debug: true
    };
};
```

I don't think the base/standard webpack config is relevant here. Let me know if you need it (but it is highly customized to an actual large-scale project).

### Code editor or IDE name and version

not relevant (wallaby core issue). Personally using it in current VS Code.

### OS name and version

not relevant (wallaby core issue). Personally using it on Windows 10 Pro (1803)

**Comment by smcenlly** [comment]
This is a duplicate of https://github.com/wallabyjs/public/issues/156. We have now added support for preserving comments in wallaby's instrumented code are are closing out both issues.

This is available in the latest version of Wallaby.js, `Wallaby.js core v1.0.629`. Visit [our docs to read more about the preserveComments setting](https://wallabyjs.com/docs/config/overview.html#preserve-comments-setting).

To address your comments above... 

The instrumentation provided by Wallaby, **and any other tool that provides test coverage reporting** is **not** the same code that will run in your production environment. While we've added support for preserving code comments, when another tool/library relies on the "shape" of your code for its own logic, it's possible that our transformations and additions will break that logic. Unfortunately, there's very little we can do if/when this breaks. We'd appreciate knowing whether support for preserving comments allows `webpack-angular-translate` library to work with wallaby.

We encourage our users to configure wallaby.js to disable as many logic/checks that aren't strictly for running your test logic. As an example, for TypeScript, wallaby does not perform type checking. However, when you use your editor or when your project compiles, the type checking is performed. By minimizing the number of unnecessary tasks that wallaby performs, you'll speed up the feedback loop provided by wallaby. In the case of your linter, you probably don't want wallaby to perform linter checks with every key-stroke, it's OK for your editor to provide you with feedback every 5-10 seconds after a change vs. in real-time for your wallaby tests. During your build/production release, we do recommend re-running all of the same tests that wallaby ran for you (as well as any linters, etc.) to ensure that you release with the same quality checks as development time.

**Comment by sambernet** [comment]
> This is a duplicate of #156. We have now added support for preserving comments in wallaby's instrumented code are are closing out both issues.

Sorry about filing a duplicate, obviously I didn't do my research properly here 🤦‍♂️ 

Thank you for still providing a detailed answer with some background considerations:

> The instrumentation provided by Wallaby, **and any other tool that provides test coverage reporting** is **not** the same code that will run in your production environment. While we've added support for preserving code comments, when another tool/library relies on the "shape" of your code for its own logic, it's possible that our transformations and additions will break that logic. Unfortunately, there's very little we can do if/when this breaks. We'd appreciate knowing whether support for preserving comments allows `webpack-angular-translate` library to work with wallaby.

We'll test this today and let you know, chances are high it will work just fine.

The `webpack-angular-translate` basically acts as a preprocessor, that removes calls that manually register i18n labels from the code (assembling the labels along the way), before passing it on to the next loader. If the calls are left in (the preprocessor is removed from wallaby config), this inevitably leads to compile errors. In fact wallaby still _seemed_ to work fine anyway, but there is no way to get rid of the error messages (besides workarounds, like implementing a wallaby preprocessor to do the same...). This was distracting and always left the feeling that wallaby _might not provide full disclosure on the state of tests_. We're thus very happy to see the option being added 😃 

> We encourage our users to configure wallaby.js to disable as many logic/checks that aren't strictly for running your test logic. As an example, for TypeScript, wallaby does not perform type checking. However, when you use your editor or when your project compiles, the type checking is performed. By minimizing the number of unnecessary tasks that wallaby performs, you'll speed up the feedback loop provided by wallaby. In the case of your linter, you probably don't want wallaby to perform linter checks with every key-stroke, it's OK for your editor to provide you with feedback every 5-10 seconds after a change vs. in real-time for your wallaby tests. During your build/production release, we do recommend re-running all of the same tests that wallaby ran for you (as well as any linters, etc.) to ensure that you release with the same quality checks as development time.

I agree that we want the feedback loop to be as fast as possible. Wallaby is doing an amazing task here - we use it on modules with 1k+ tests and it mostly provides instant feedback on the tests. For your example, I would actually be willing to take some moderate hit on the feedback time when it would buy me TS type-checking within that feedback loop. It allows me to ensure my code is also formally correct, without having to run a regular build also at the end of a TDD session to ensure I'm really in a "green" state. Especially as typechecking can (in theory) be done asynchronously *after* the test feedback has already been provided.

But this is again one special case. As always in software engineering, disabling logic/checks in the wallaby-specific webpack config is about trade-offs. And I agree that the primary concern of this tool is to give test feedback as quick as possible.

**Comment by sambernet** [comment]
I can confirm that the new setting fixes our issue for good. 👍 

Thanks guys!
